### ISSUE TITLE
Maintain several "computation graphs"

### ISSUE THREAD

**Issue by anhmeow** [pytorch/xla#2065]
I understand that the TPU need to "create/build" a computation graph and this is the reason why first batch takes way longer than other batches => That's fair enough.

I also understand that if the computation required is changing (like changing the batch size), the TPU need to "create/build" a new graph and, so, the first batch with the change will take some times => That's also ok for me.

What i don't understand is that : if i move from a specific computation "type A" to a computation "type B", when i go back to "type A", seems that the computation graph is rebuild/recomputed.

As comparison with tensorflow (disclaimer : i'm not comparing the implementation or library just  the behavior ;) ), once a graph has been built for a type of computation, seems like it's kept and can be reuse.

Put in context, for exemple, when i train a model, and performing validation at epoch end, i have, on colab TPU : 
- wait like 2 minutes to start the train (first batch) => ok, i inderstand
- when validation is launch, wait like 2 minutes too => ok, i understand
- when i proceed new epoch and train, wait again like 2 minutes => some optimization should exist, no?

If i remove the validation, i wait 2 minutes only at the first batch of the first epoch, that's all.

Is there any way to solve this?

**Comment by dlibenzi** [comment]
There are many graphs.
The `PyTorch -> IR` graph is always happening.
The `IR -> HLO` is only happening if the IR graph changed.
The `HLO -> Binary` (XLA compilation) only happens if the previous happen. Actually, since on TPU there is a persistent (until you restart the TPU) cache on the TPU VM, that might not happen at all.

So if you really have the same TypeA graph at time T0 and T2, the latter should just be cached and not even trigger the `IR -> HLO` lowering.

**Comment by anhmeow** [comment]
Thanks a lot for the explanations !

Is there a way i can check/logs if the IR -> HLO is happening to confirm if i'm in this situation or not ?

**Comment by dlibenzi** [comment]
Probably the best way is to use debug_run and let the model run about 10 steps, then hit CTRL^C and look at the generated data.

https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#using-debug_runpy-to-collect-debug-information

**Comment by Jeevesh8** [comment]
Would the graph be different, if there are if-else conditions/loops/different sequence length in case of RNNs ? I think IRs can represent different if-else conditions and for-loops.. but I am not sure.

**Comment by Jeevesh8** [comment]
Also, I read that computation graph must be exactly same across all TPU cores. So does that mean that XLA pads all 'sequence_lengths' (as in RNN) across all samples across all cores to same length ?

**Comment by dlibenzi** [comment]
This is an example I did some time ago, of an LSTM skipping padded values.
It uses torch.where(), which does not create different graphs:

https://gist.github.com/dlibenzi/d049894d6e1cde15f0d39a6149a1fdb5#file-xla_lstm-py-L56

**Comment by Jeevesh8** [comment]
The example is nice. I understand that the graph will be same in case of loops, if-else conditions inside models.  But, had we used direct .bmm() multiplication(instead of for-loop) for multiplying <seq_len> (where <seq_len> is variable) number of matrices, like in a transformer, the compilation would have repeated, unless we make sure that each batch of transformer is padded to same length.  Am I right ?
Also, your implementation would take less memory as using torch.where() prevents padding values to add lots of new parts in computation graph.

**Comment by dlibenzi** [comment]
Yes, if you have variable seqlen where you stop the loop at the effective size, you will have as many graphs as the length of the unique set of seqlens.
Padding adds a cost, as now you have to waste FLOPS to process the padding.
This is why one creates buckets (3..4 usually), and not a single one. To minimize the wasted FLOPS.

**Comment by Jeevesh8** [comment]
What are buckets ? Like GCP buckets ?
